decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,485 @@
+
+### Bindings
+
+Research has shown that the common denominator when addressing the above
+listed languages is a `C` interface. A standard `C` API surface, with
+clear and manageable ownership semantics would facilitate support in the
+broadest range of future programming environments.
+
+Ideally, any bindings would also allow idiomatic programming within
+the target language, rather than forcing specific design patterns from
+any internals of OpenAssetIO - such as synchronous programming, RAII,
+etc.
+
+### Performance
+
+A core requirement raised by numerous parties is that the "happy path"
+should be as performant as possible. Thus, it is required that any
+mechanism introduced to broaden compatibility doesn't compromise the
+(current) majority deployment scenarios.
+
+Of note, a `C++` host talking to a `C++` manager has been the most
+frequently requested high-performance pathway.
+
+### Development lifecycle
+
+The overhead of maintaining dynamically linked `C++` libraries within a
+plugin-based architecture is often fraught with constantly chasing ABI
+changes and/or conflicts between mutual dependencies.
+
+Foundry has prior art in using `C` function pointer suites as a compiler
+isolation technique on top of `C++` implementations that can help
+de-couple the plugin host and the specifics of any given plugin.
+
+In the constantly evolving landscape defining which dependency versions
+are in use within which versions of common DCC tools, avoiding the need
+to matrix-compile and juggle plugin search paths per-host would be
+massively beneficial.
+
+## Options considered
+
+### Option 1
+
+`C++` and `Python` support only.
+
+**Estimated Cost**: Small
+
+#### Pros
+
+-   Simplifies development.
+
+#### Cons
+
+-   Leaves use within languages other than these a potentially
+    challenging exercise left to the reader.
+-   Compiler isolation support would need to be handled as an add-on
+    through an additional layer such as a `C` suite, `protobuf` or
+    similar.
+
+### Option 2
+
+`C` core library, bound to `C++` and `Python` as appropriate.
+
+**Estimated Cost**: Large
+
+#### Pros
+
+-   Makes future bindings or adoption in `C`-based languages far
+    simpler as all core API functionality is inherently available
+    through idiomatic `C`.
+-   Inherently presents a stable ABI isolation layer between hosts and
+    plugins.
+
+#### Cons
+
+-   The data model of OpenAssetIO _requires_ `map` and `optional` types.
+    The lack of standardization in `C` around these concepts means a
+    local implementation of these would be required. This, and their
+    mapping to C++ and Python introduces run-time and maintenance
+    overhead, as well as increased code complexity and project risk.
+-   `C` is notoriously unsafe with regards to type-safety and memory
+    management. Higher-level systems programming languages, such as
+    `C++`
+    and Rust, add well-known idioms and/or protective mechanisms (both
```

Comment:
[fix] Elsewhere programming languages are surrounded in backticks, but not "Rust" here.

Reply:
Should be consistent [now](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/b829e705dbea283d8ffd98f06294a71ffc031bd5).